Add javaagent<->application context bridge for HttpRouteHolder

mateuszrzeszutek opened the pull request:
Resolves #4319 -- there are no more context keys worth bridging in both instrumentation-api and instrumentation-api-semconv, this was the last useful one.

trask commented on instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpRouteHolder.java:
```diff
@@ -151,10 +156,38 @@ private boolean isBetterRoute(String name) {
     return httpRouteHolder == null ? null : httpRouteHolder.route;
   }
 
+  /**
+   * This method is internal and is hence not for public use. Its API is unstable and can change at
+   * any time.
+   */
```
I don't think these disclaimers are needed (guessing they were originally public before changing access to reflection?)

trask replied:
wait, how is the MethodHandle able to access this? 🤔

mateuszrzeszutek replied:
Removed the disclaimer comments.

trask reviewed (approved):
🎉

trask commented on instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java:
```diff
@@ -0,0 +1,171 @@
+      applicationGetRoute =
+          lookup.findVirtual(
+              applicationHttpRouteHolder, "internalGetRoute", MethodType.methodType(String.class));
+    } catch (Throwable ignored) {
```
I think something's not quite right. I added `printStackTrace` here and got:

```
java.lang.IllegalAccessException: no such method: io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpRouteHolder.internalCreate(int,String)HttpRouteHolder/invokeStatic
	at java.base/java.lang.invoke.MemberName.makeAccessException(MemberName.java:972)
	at java.base/java.lang.invoke.MemberName$Factory.resolveOrFail(MemberName.java:1117)
	at java.base/java.lang.invoke.MethodHandles$Lookup.resolveOrFail(MethodHandles.java:3649)
	at java.base/java.lang.invoke.MethodHandles$Lookup.findStatic(MethodHandles.java:2588)
	at io.opentelemetry.javaagent.instrumentation.opentelemetryapi.context.InstrumentationApiContextBridging.<clinit>(InstrumentationApiContextBridging.java:91)
```

mateuszrzeszutek replied:
Now it's all clear: the `lookup()` method is not actually magic and it doesn't magically make things work 😂 

Turns out that the test was incomplete, and only verified one particular scenario (our of many possible).

trask commented on instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/context/InstrumentationApiContextBridging.java:
```diff
@@ -0,0 +1,171 @@
+
+    ContextKeyBridge<?, ?> httpRouteHolderBridge = httpRouteHolderBridge();
+    if (httpRouteHolderBridge != null) {
+      bridges.add(httpRouteHolderBridge);
```
and if I remove the http route bridge here, the tests still pass

mateuszrzeszutek replied:
Right, the test wasn't properly constructed... fixed that.
Thanks!

mateuszrzeszutek commented on instrumentation/opentelemetry-api/opentelemetry-api-1.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/opentelemetryapi/OpenTelemetryApiInstrumentationModule.java:
```diff
@@ -24,6 +25,8 @@ public List<TypeInstrumentation> typeInstrumentations() {
         new OpenTelemetryInstrumentation(),
-        new SpanInstrumentation());
+        new SpanInstrumentation(),
+        // instrumentation-api specific instrumentation
+        new HttpRouteStateInstrumentation());
```
The solution to this issue turned out to be a bit more complex than expected -- I don't particularly like this, but I had to insert additional instrumentation-api instrumentation into the opentelemetry-api module, cause it uses the `AgentContextWrapper`. It doesn't add any muzzle dependencies though, so it should be completely neutral for applications that just use opentelemetry-api without instrumentation-api.

trask commented:
I haven't thought this through, so maybe missing something, but wondering if we can instead instrument `HttpRouteHolder.get()` and `HttpRouteHolder.updateRoute()`, we may not need a context key bridge for this, and then the instrumentation could live outside of the opentelemetry-api bridge

trask commented:
(a bit like #4786)

mateuszrzeszutek commented:
> I haven't thought this through, so maybe missing something, but wondering if we can instead instrument `HttpRouteHolder.get()` and `HttpRouteHolder.updateRoute()`, we may not need a context key bridge for this, and then the instrumentation could live outside of the opentelemetry-api bridge

I thought that this would be a bit more difficult, since we'd have to bridge `HttpRouteSource` and `HttpRouteBiGetter` too, not just the `Context`. I left the context key bridging and extracted a separate instrumentation-api instrumentation (like https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4786); I also moved all the instrumentation-api context bridging tests there.

trask reviewed (approved):
thx, I have a couple more thoughts about bridging, will explore in a follow-up 👍
